Platform/TigerlakeBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c

```diff
@@ -58,6 +58,112 @@ GetBoardId (
   OUT UINT8 *BoardId
 );
 
+/* BAD DSO mark */
+CONST UINT64 mMagicBadTcct = SIGNATURE_64('B', 'A', 'D', '_', 'T', 'C', 'C', 'T');
+
+/**
+  Check if the BAD DSO mark exist or not
+
+  @retval TRUE  DSO was marked as BAD DSO
+  @retval FALSE No BAD DSO mark found
+
+ */
+BOOLEAN
+EFIAPI
+IsMarkedBadDso (
+  VOID
+)
+{
+  EFI_STATUS  Status;
+  UINT32      Address;
+  UINT64      Data;
+
+  CONTAINER_ENTRY  *ContainerEntry;
+  COMPONENT_ENTRY  *CompEntry;
+  CONTAINER_HDR    *ContainerHdr;
+
+  DEBUG ((DEBUG_INFO, "Check Bad DSO mark\n"));
+
+  Status = LocateComponentEntry (SIGNATURE_32 ('I', 'P', 'F', 'W'),
+                                 SIGNATURE_32 ('T', 'C', 'C', 'T'),
+                                 &ContainerEntry, &CompEntry);
+  if (EFI_ERROR (Status) || (ContainerEntry == NULL) || (CompEntry == NULL)) {
+    return FALSE;
+  }
+
+  ContainerHdr = (CONTAINER_HDR *)(UINTN)ContainerEntry->HeaderCache;
+
+  Address = ContainerEntry->Base + ContainerHdr->DataOffset + CompEntry->Offset;
+  Data = *(UINT64 *)Address;
+
+  if (Data == mMagicBadTcct) {
+    DEBUG ((DEBUG_INFO, "Detect BAD DSO(TCCT) mark!\n"));
+    return TRUE;
+  } else {
+    return FALSE;
+  }
+}
+
+/**
+  Invalidate Bad DSO
+
+  @retval EFI_SUCCESS
+  @retval EFI_NOT_FOUND          Unable to find IPFW/TCCT
+  @retval EFI_OUT_OF_RESOURCES   Bios region is too small
+  @retval Others                 Errors during SPI operations
+
+ */
+EFI_STATUS
+EFIAPI
+InvalidateBadDso (
+  VOID
+)
+{
+  EFI_STATUS  Status;
+  UINT32      Address;
+  UINT32      Length;
+  UINT32      BaseAddress;
+  UINT32      RegionSize;
+  CONTAINER_ENTRY  *ContainerEntry;
+  COMPONENT_ENTRY  *CompEntry;
+  CONTAINER_HDR    *ContainerHdr;
+
+  DEBUG ((DEBUG_INFO, "Invalidate Bad DSO\n"));
+
+  Status = LocateComponentEntry (SIGNATURE_32 ('I', 'P', 'F', 'W'),
+                                 SIGNATURE_32 ('T', 'C', 'C', 'T'),
+                                 &ContainerEntry, &CompEntry);
+  if (EFI_ERROR (Status) || (ContainerEntry == NULL) || (CompEntry == NULL)) {
+    return EFI_NOT_FOUND;
+  }
+  ContainerHdr = (CONTAINER_HDR *)(UINTN)ContainerEntry->HeaderCache;
+
+  Address = ContainerEntry->Base + ContainerHdr->DataOffset + CompEntry->Offset;
+  Length = KB_(4); /* minimum erase size */
+
+  Status = SpiGetRegionAddress (FlashRegionBios, &BaseAddress, &RegionSize);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  BaseAddress = ((UINT32)(~RegionSize) + 1);
+
+  Address -= BaseAddress;
+  if ((Address + Length) > RegionSize) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  Status = SpiFlashErase (FlashRegionBios, Address, Length);
```

Comment:
Can we just write 0 to the location as a bad marker ?  In this way, the erasing can be avoided to reduce flash erasing frequency.  For flash, you can write from 1 to 0 without erasing.  But to write from 0 to 1, erasing is required. So writing 0 does not need any erasing.

Reply:
Good suggestion. Thanks. Will update accordingly.

Reply:
Unfortunately, all-zero Component will break anti-rollback, so my new patch keeps "Svn" field non-zero.